v2/spiffeid/path.go

```diff
@@ -0,0 +1,106 @@
+package spiffeid
+
+import (
+	"bytes"
+	"fmt"
+)
+
+// ValidatePath validates that a path string is a conformant path for a SPIFFE
+// ID. Namely:
+// - does not contain an empty segments (including a trailing slash)
+// - does not contain dot segments (i.e. '.' or '..')
+// - does not contain any percent encoded characters
+// - has only characters from the unreserved or sub-delims set from RFC3986.
+func ValidatePath(path string) error {
+	if reason := validatePath(path); reason != noReason {
+		return pathErr{path: path, reason: reason}
+	}
+	return nil
+}
+
+func validatePath(path string) errReason {
+	switch {
+	case path == "":
+		return noReason
+	case path[0] != '/':
+		return noLeadingSlash
+	}
+
+	segmentStart := 0
+	segmentEnd := 0
+	for ; segmentEnd < len(path); segmentEnd++ {
+		c := path[segmentEnd]
+		if c == '/' {
+			switch path[segmentStart:segmentEnd] {
+			case "/":
+				return emptySegment
+			case "/.", "/..":
+				return dotSegment
+			}
+			segmentStart = segmentEnd
+			continue
+		}
+		if !isValidPathSegmentChar(c) {
+			return badPathSegmentChar
+		}
+	}
+
+	switch path[segmentStart:segmentEnd] {
+	case "/":
+		return trailingSlash
+	case "/.", "/..":
+		return dotSegment
+	}
+	return noReason
+}
+
+// isValidPathSegmentChar returns true if the given character is acceptable in
+func formatPath(format string, args ...interface{}) (string, error) {
+	path := fmt.Sprintf(format, args...)
+	if err := ValidatePath(path); err != nil {
+		return "", err
+	}
+	return path, nil
+}
+
+func joinPathSegments(segments ...string) (string, error) {
+	buf := new(bytes.Buffer)
+	for _, segment := range segments {
+		if err := validatePathSegment(segment); err != nil {
+			return "", err
+		}
+		buf.WriteByte('/')
+		buf.WriteString(segment)
+	}
+	return buf.String(), nil
+}
+
+func validatePathSegment(segment string) error {
+	if segment == "" {
+		return segmentErr{segment: segment, reason: empty}
+	}
+	for i := 0; i < len(segment); i++ {
+		if !isValidPathSegmentChar(segment[i]) {
+			return segmentErr{segment: segment, reason: badPathSegmentChar}
+		}
+	}
+	return nil
+}
+
+// a path segment. Acceptable path segment characters consist of the the
```

Comment:
This comment got split with the one that remained on formatPath. I've removed it since it isn't really adding anything that needs clarification and is on an unexported method.